docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,826 @@
+# Rework of Mesh*Routes matching
+
+* Status: accepted
+
+Technical story: https://github.com/kumahq/kuma/issues/10247
+
+## Context and Problem Statement
+
+At the moment when targeting `Mesh*Route` you need to put it in topLevel targetRef. 
+We want to iterate over policy design and move `Mesh*Route` to `spec.to[].targetRef` section follow model where `spec.targetRef`
+selects whole proxies, `spec.from[].targetRef` selects inbounds and `spec.to[].targetRef` selects outbounds
+
+## Decision Outcome
+
+- add possibility to put `Mesh*Routes` in `spec.to[].targetRef`
+
+## Considered options
+
+1. add possibility to put `Mesh*Routes` in `spec.to[].targetRef` -> `spec.targetRef` selects proxies, `spec.from[].targetRef` selects inbounds, `spec.to[].targetRef` selects outbounds.
+2. leave `Mesh*Routes` in top level targetRef as it is now, improve the semantics of `spec.targetRef` and introduce `spec.workloadSelector` to select whole proxies
+
+### Add possibility to put Mesh*Routes in `spec.to[].targetRef`
+
+#### Advantages
+
+- simplifying policy api as `spec.targetRef` will always pick proxy and `spec.to[].targetRef` section will select outgoing traffic
+- ability to set different confs on different proxy for the same route (overriding default conf for route without producer/consumer policies, on universal)
+- gives more flexibility as you can target multiple outbounds in single policy. Reducing number of consumer policies
+- more consistent with current approach
+- easier policy matching code
+- least problematic migration
+
+#### Disadvantages
+
+- can be harder to understand because `spec.targetRef` can be MeshSubset (which is optional and probably will rarely be used)
+- yet another migration for users (only for system policies as for namespaced policies this is already not allowed) 
+- mixing outbounds and routes in `spec.to[].targetRef`
+- hides the most often most important information under `to[]` adding noise when targeting Mesh in `spec.targetRef`
+- is it consistent in the case of MeshGateways? MeshHTTPRoutes aren't on outbound listeners in the case of MeshGateways, they're on inbound listeners
```

Comment:
Do we care? `To` / `from` don't have to always map to `outbound` / `inbound`. A route always routes "to" something right?

Reply:
Sure the route routes "to" something but that's not what `to` is referring to, it's targetting traffic "to" the route. I.e. for a normal mesh service, when we have `.to[].targetRef.kind: MeshService` a policy targets traffic "to" a service, when we have `.to[].targetRef.kind: MeshHTTPRoute` it targets traffic "to" the route. But it doesn't make sense for a gateway to say a policy targets traffic "to" a route, the traffic is going "to" a gateway listener but the routes determine where traffic goes "to". A gateway listener is an _inbound_ listener.

The point is that it's not really useful or meaningful to talk about "to" or "from" a route, it's not useful to _also_ have this on the policy. The route resource itself determines whether it's "from" or "to". We're just targeting the route.

And my claim and motivation for this variant is that it's not even useful _ever_ to have "to" and "from". Given a policy kind `SomePolicy` and a targeted kind `Target`, you can never have both `to[].kind.targetRef.kind: Target` and `from[].kind.targetRef.kind: Target`. The semantics of `SomePolicy` and `Target` uniquely determine whether you use `to` or `from`, so you might as well have `spec.targetRef.kind: Target`.

Reply:
Is it that big of the noise to justify yet another big migration?
I'm not even talking about code, I'm talking about "docs", "explaining it to users", "deprecating", asking for a change etc.

We are talking about
```
spec:
  targetRef:
    kind: MeshService
    name: backend
```
vs
```
spec:
  to:
  - targetRef:
      kind: MeshService
      name: backend
```

How is it really better than the second option? What's the value here?

I'm reading advantages

> - the workloads targeted by the policy is often "all workloads" or can be
  inferred from the namespace so we kind of waste the top level, "special" `spec.targetRef`

what does "waste" mean here? I don't think I get this argument

> tends to be less noisy as we move most interesting part of the policy from `to` /`from`section

are users confused by `to`? Is it difficult to construct the policies with `to` and it will be much easier if we were to go with the first option?

> - Doesn't force every potential resource into a "to" or a "from", instead using the semantics of the `kind` itself
	- `HTTPRoute` is a route, regardless of whether it's inbound or outbound

but HTTPRoute is "outbound" route, no? Wouldn't we have a separate resource for inbound route?

> WorkloadIdentity

left the comment in another place about it.

I miss inbound policies in this discussion.

If we were starting this from scratch maybe that would be a _slightly_ better way, but is it worth the effort now?

Reply:
> Wouldn't we have a separate resource for inbound route?

We have the resource on gateways where it definitely doesn't fit the definition of "outbound route".

No I don't think it's worth the effort, I just want it documented and the drawbacks and implications of the current method understood.

Reply:
> what does "waste" mean here? I don't think I get this argument

In the context of gateway API spec.targetRef has a special meaning. Our API relegates this to "almost never used" and adds a _different_ targetRef to "100% always used". Thus we have wasted spec.targetRef